### PR TITLE
US47787-Included OS service image resource and update existing resource with new client

### PR DESCRIPTION
**Pull request by dbozzato81**
This PR includes: 

- New `OS service image` resource and relevant life cycle
- Updated metal client dependency

### CODE REVIEW

File: examples/resources/hpegl_metal_image/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+hpegl_metal_image.terra_image: Creating...
+hpegl_metal_image.terra_image: Creation complete after 2s [id=3de686c3-654b-4613-92b5-407f77b2dd13]
+
+Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
+
+```
+
+### Argument Reference
+
+The available_resources block takes the following arguments.
```
Comment (anurag-shrivastav) [inline examples/resources/hpegl_metal_image/README.md:42]: 'available_resources' or 'resource'?

File: examples/resources/hpegl_metal_image/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+- `os_service_image_file` - Path to the YAML file containig the OS service image definition.
+
+### Attribute Reference
+
+There are not additinoal attributes.
```
Comment (anurag-shrivastav) [inline examples/resources/hpegl_metal_image/README.md:48]:
```suggestion
There are no additional attributes.
```

File: examples/resources/hpegl_metal_image/README.md (hunk 1 above)
Comment (dbozzato81) [reply examples/resources/hpegl_metal_image/README.md:42]: Totally resource thanks!

File: examples/README.md [hunk 3]
```diff
@@ -113,12 +113,13 @@ output "tfhost" {
...
-1. [Volume creation](./volume-creation/README.md): Create iSCSI volumes for host attachments.
-1. [Network creation](./network-creation/README.md): Create custom new metworks for project intra-communication.
-1. [Usage information](./usage/README.md): Extract resource usage information.
-1. [Project](./project/README.md): Create and manipulate projects.
+1. [Available resrources](./data-sources/hpegl_metal_available_resources/README.md): Obtain information about unprovisioned resources available to terraform.
+1. [Available images](./data-sources/hpegl_metal_available_images/README.md): Obtain filtered, specifc image information.
+1. [SSH key creation](./resources/hpegl_metal_ssh_key/README.md): Create new SSH keys for host image injection.
+1. [Host creation](./resources/hpegl_metal_host/README.md): Create one or more hosts.
+1. [Volume creation](./resources/hpegl_metal_volume/README.md): Create iSCSI volumes for host attachments.
+1. [Network creation](./resources/hpegl_metal_network/README.md): Create custom new metworks for project intra-communication.
```
Comment (mchuang3) [inline examples/README.md:121]:
```suggestion
1. [Network creation](./resources/hpegl_metal_network/README.md): Create custom new networks for intra-project communication.
```

File: examples/README.md [hunk 4]
```diff
@@ -113,12 +113,13 @@ output "tfhost" {
...
-1. [Host creation](./host-creation/README.md): Create one or more hosts.
-1. [Volume creation](./volume-creation/README.md): Create iSCSI volumes for host attachments.
-1. [Network creation](./network-creation/README.md): Create custom new metworks for project intra-communication.
-1. [Usage information](./usage/README.md): Extract resource usage information.
-1. [Project](./project/README.md): Create and manipulate projects.
+1. [Available resrources](./data-sources/hpegl_metal_available_resources/README.md): Obtain information about unprovisioned resources available to terraform.
+1. [Available images](./data-sources/hpegl_metal_available_images/README.md): Obtain filtered, specifc image information.
+1. [SSH key creation](./resources/hpegl_metal_ssh_key/README.md): Create new SSH keys for host image injection.
+1. [Host creation](./resources/hpegl_metal_host/README.md): Create one or more hosts.
+1. [Volume creation](./resources/hpegl_metal_volume/README.md): Create iSCSI volumes for host attachments.
```
Comment (mchuang3) [inline examples/README.md:120]: We will have FC volumes soon, so keep this generic.
```suggestion
1. [Volume creation](./resources/hpegl_metal_volume/README.md): Create storage volumes for host attachments.
```

File: examples/resources/hpegl_metal_image/service.yml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  svc_ver: 18.04-20201103
+  description: v2 service client install service test
+  hoster_use: true
+  project_use: true
+  image_info:
+    image_size: 0
+    urls: []
+    download_timeout: 0
+    signature: ""
+    algorithm: sha256sum
```
Comment (mchuang3) [inline examples/resources/hpegl_metal_image/service.yml:19]: This is deprecated. See quake/model/service.go. Remove it from example.
```suggestion
```

File: examples/resources/hpegl_metal_project/README.md [hunk 6]
```diff
@@ -69,6 +69,7 @@ The available_resources block takes the following arguments.
   - `volume_capacity` - Maximum aggregated volume capacity in TiB
   - `private_networks` - Maximum number of private networks
   - `instance_types` - (Optional) Map of instance type ID to maximum number of hosts that can be created with that instance type
+- `permitted_images` - (Optional) List of OS service images ID allowed to be used by compute instance of this compute group
```
Comment (mchuang3) [inline examples/resources/hpegl_metal_project/README.md:72]:
```suggestion
- `permitted_images` - (Optional) List of OS service image IDs allowed to be used by compute instances of this compute group
```

File: internal/acceptance_test/resource_image_test.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+
+	"github.com/hewlettpackard/hpegl-metal-terraform-resources/pkg/client"
+)
+
+func TestAccResourceImage_Basic(t *testing.T) {
+	// as-of Project creation is only supported when using GL IAM token.
```
Comment (mchuang3) [inline internal/acceptance_test/resource_image_test.go:17]: ? As of now, Project creation...?

File: internal/resources/resource_image.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	return &schema.Resource{
+		Create: resourceMetalImageCreate,
+		Read:   resourceMetalImageRead,
+		Delete: resourceMetalImageDelete,
+		Update: resourceMetalImageUpdate,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema:      serviceSchema(),
+		Description: "Provides service image resource. This allows creation, deletion and update of Metal OS service images .",
```
Comment (mchuang3) [inline internal/resources/resource_image.go:36]:
```suggestion
		Description: "Provides service image resource. This allows creation, deletion and update of Metal OS service images.",
```

File: internal/resources/resource_image.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+func resourceMetalImageCreate(d *schema.ResourceData, meta interface{}) (err error) {
+	defer wrapResourceError(&err, "failed to create OS service image")
+
+	filePath := safeString(d.Get(iServiceImageFile))
+
+	file, err := os.Open(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to open file %s, %v", filePath, err)
```
Comment (mchuang3) [inline internal/resources/resource_image.go:47]:
```suggestion
		return fmt.Errorf("open file %s, %v", filePath, err)
```

File: internal/resources/resource_image.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	p, err := client.GetClientFromMetaMap(meta)
+	if err != nil {
+		return fmt.Errorf("failed to get client, %v", err)
```
Comment (mchuang3) [inline internal/resources/resource_image.go:52]: Same thing else where in this PR. Remove the "failed to..." in returned error.
```suggestion
		return fmt.Errorf("get client, %v", err)
```

File: internal/acceptance_test/resource_image_test.go (hunk 7 above)
Comment (dbozzato81) [reply internal/acceptance_test/resource_image_test.go:17]: ops, fixed

**Review by mchuang3 — APPROVED** [review]

**Pull request merged**